ctru-rs/src/services/soc.rs

```diff
@@ -31,20 +31,28 @@ impl Soc {
     ///
     /// This function will return an error if the `Soc` service is already initialized
     pub fn init_with_buffer_size(num_bytes: usize) -> crate::Result<Self> {
-        match SOC_ACTIVE.compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst) {
-            Ok(_) => unsafe {
-                let soc_mem = memalign(0x1000, num_bytes) as *mut u32;
-
-                let r = socInit(soc_mem, num_bytes as u32);
+        let _service_handler = ServiceHandler::new(
+            &SOC_ACTIVE,
+            true,
+            || {
+                let soc_mem = unsafe { memalign(0x1000, num_bytes) } as *mut u32;
+                let r = unsafe { ctru_sys::socInit(soc_mem, num_bytes as u32) };
                 if r < 0 {
-                    free(soc_mem as *mut _);
-                    Err(r.into())
-                } else {
-                    Ok(Self { soc_mem })
+                    return Err(r.into());
                 }
+
+                Ok(())
+            },
+            // `socExit` returns an error code. There is no documentantion of when errors could happen,
+            // but we wouldn't be able to handle them in the `Drop` implementation anyways.
+            // Surely nothing bad will happens :D
+            || unsafe {
+                // The socket buffer is freed automatically by `socExit`
+                ctru_sys::socExit();
```

Comment:
Is this true? I couldn't find the place it's freed in libctru. Also, does that mean that the previous code was doing a double free?

Reply:
While searching around for this buffer, I saw this note in the docs:

> Note: The specified context buffer can no longer be accessed by the process which called this function, since the userland permissions for this block are set to no-access.

Asking in the homebrew forums for confirmation, Fenrir himself responded telling me a free is probably the case, possibly handled by the OS itself when the process is closed. A double free panic didn’t happen possibly because the functions (`free` has no way of error handling) simply ignore OS errors.